File: php/src/Advertisement/Domain/ValueObject/Password.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Domain\ValueObject;


use Exception;

final readonly class Password
{
    private function __construct(private string $value)
    {
    }

    /**
     * @throws Exception
     */
    public static function fromPlainPassword(string $password): self
    {
        $result = password_hash($password, PASSWORD_ARGON2I);
        if(null === $result || false === $result) {
            throw new Exception("Problem hashing password");
        }
        return new Password($result);
    }

    public function value(): string
    {
        return $this->value;
    }

    public static function fromEncryptedPassword(string $password): self
    {
        return new Password($password);
    }

    public function isValidatedWith(string $password): bool
    {
        $hashSpecs = password_get_info($this->value);

        if(null === $hashSpecs['algo']) {
            return $this->value === md5($password);
        }

        return password_verify($password, $this->value);
    }
}

File: kotlin/src/main/kotlin/advertisement/domain/model/value_object/Password.kt
package advertisement.domain.model.value_object

import advertisement.infrastructure.PasswordHasher

class Password private constructor(private val value: String) {

    companion object {
        fun fromPlainPassword(password: String, passwordHasher: PasswordHasher): Password {
            val encryptedPassword = passwordHasher.create(password)

            return Password(encryptedPassword)
        }

        fun fromEncryptedPassword(encryptedPassword: String): Password {
            return Password(encryptedPassword)
        }
    }

    fun value(): String {
        return value
    }

    fun validateWith(password: String, passwordHasher: PasswordHasher): Boolean {
        return passwordHasher.verify(value, password)
    }
}
